# 📦 Amazon India Products — Exploratory Data Analysis
**CodeAlpha Data Analytics Internship | Task 2**

| | |
|---|---|
| **Dataset** | Amazon India Product Reviews (1,465 products) |
| **Task** | Task 2 — Exploratory Data Analysis |
| **Tools** | Python · Pandas · NumPy |
---


In [1]:
import pandas as pd
import numpy as np
import warnings
warnings.filterwarnings('ignore')
print("✅ Libraries loaded")


✅ Libraries loaded


## 1. Load Dataset

In [2]:
df_raw = pd.read_csv('../amazon.csv')
print(f"Raw shape: {df_raw.shape}")
df_raw.head(3)


Raw shape: (1465, 16)


,product_id,product_name,category,discounted_price,actual_price,discount_percentage,rating,rating_count,about_product,user_id,user_name,review_id,review_title,review_content,img_link,product_link
0,B07JW9H4J1,Wayona Nylon Braided USB to Lightning Fast Cha...,Computers&Accessories|Accessories&Peripherals|...,₹399,"₹1,099",64%,4.2,"24,269",High Compatibility : Compatible With iPhone 12...,"AG3D6O4STAQKAY2UVGEUV46KN35Q,AHMY5CWJMMK5BJRBB...","Manav,Adarsh gupta,Sundeep,S.Sayeed Ahmed,jasp...","R3HXWT0LRP0NMF,R2AJM3LFTLZHFO,R6AQJGUP6P86,R1K...","Satisfied,Charging is really fast,Value for mo...",Looks durable Charging is fine tooNo complains...,https://m.media-amazon.com/images/W/WEBP_40237...,https://www.amazon.in/Wayona-Braided-WN3LG1-Sy...
1,B098NS6PVG,Ambrane Unbreakable 60W / 3A Fast Charging 1.5...,Computers&Accessories|Accessories&Peripherals|...,₹199,₹349,43%,4.0,"43,994","Compatible with all Type C enabled devices, be...","AECPFYFQVRUWC3KGNLJIOREFP5LQ,AGYYVPDD7YG7FYNBX...","ArdKn,Nirbhay kumar,Sagar Viswanathan,Asp,Plac...","RGIQEG07R9HS2,R1SMWZQ86XIN8U,R2J3Y1WL29GWDE,RY...","A Good Braided Cable for Your Type C Device,Go...",I ordered this cable to connect my phone to An...,https://m.media-amazon.com/images/W/WEBP_40237...,https://www.amazon.in/Ambrane-Unbreakable-Char...
2,B096MSW6CT,Sounce Fast Phone Charging Cable & Data Sync U...,Computers&Accessories|Accessories&Peripherals|...,₹199,"₹1,899",90%,3.9,"7,928",【 Fast Charger& Data Sync】-With built-in safet...,"AGU3BBQ2V2DDAMOAKGFAWDDQ6QHA,AESFLDV2PT363T2AQ...","Kunal,Himanshu,viswanath,sai niharka,saqib mal...","R3J3EQQ9TZI5ZJ,R3E7WBGK7ID0KV,RWU79XKQ6I1QF,R2...","Good speed for earlier versions,Good Product,W...","Not quite durable and sturdy,https://m.media-a...",https://m.media-amazon.com/images/W/WEBP_40237...,https://www.amazon.in/Sounce-iPhone-Charging-C...


## 2. Data Cleaning & Type Conversion

In [3]:
df = df_raw.copy()

def clean_price(col):
    return col.str.replace('₹', '', regex=False).str.replace(',', '', regex=False).str.strip().astype(float)

df['discounted_price'] = clean_price(df['discounted_price'])
df['actual_price']     = clean_price(df['actual_price'])
df['discount_pct']     = df['discount_percentage'].str.replace('%', '', regex=False).str.strip().astype(float)
df['rating']           = pd.to_numeric(df['rating'], errors='coerce')
df['rating_count']     = df['rating_count'].str.replace(',', '', regex=False).str.strip()
df['rating_count']     = pd.to_numeric(df['rating_count'], errors='coerce')
df['main_category']    = df['category'].str.split('|').str[0]
df['sub_category']     = df['category'].str.split('|').str[1].fillna('Unknown')
df['savings']          = df['actual_price'] - df['discounted_price']

df_clean = df.dropna(subset=['rating']).copy()
print(f"After cleaning: {df_clean.shape[0]} rows  (dropped {df.shape[0]-df_clean.shape[0]} invalid rows)")
df_clean[['product_name','main_category','discounted_price','actual_price','discount_pct','rating','rating_count']].head()


After cleaning: 1464 rows  (dropped 1 invalid rows)


,product_name,main_category,discounted_price,actual_price,discount_pct,rating,rating_count
0,Wayona Nylon Braided USB to Lightning Fast Cha...,Computers&Accessories,399.0,1099.0,64.0,4.2,24269.0
1,Ambrane Unbreakable 60W / 3A Fast Charging 1.5...,Computers&Accessories,199.0,349.0,43.0,4.0,43994.0
2,Sounce Fast Phone Charging Cable & Data Sync U...,Computers&Accessories,199.0,1899.0,90.0,3.9,7928.0
3,boAt Deuce USB 300 2 in 1 Type-C & Micro USB S...,Computers&Accessories,329.0,699.0,53.0,4.2,94363.0
4,Portronics Konnect L 1.2M Fast Charging 3A 8 P...,Computers&Accessories,154.0,399.0,61.0,4.2,16905.0


## 3. Dataset Structure

In [4]:
print("="*50)
print("DATASET INFO")
print("="*50)
df_clean.info()


DATASET INFO
<class 'pandas.DataFrame'>
Index: 1464 entries, 0 to 1464
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   product_id           1464 non-null   str    
 1   product_name         1464 non-null   str    
 2   category             1464 non-null   str    
 3   discounted_price     1464 non-null   float64
 4   actual_price         1464 non-null   float64
 5   discount_percentage  1464 non-null   str    
 6   rating               1464 non-null   float64
 7   rating_count         1462 non-null   float64
 8   about_product        1464 non-null   str    
 9   user_id              1464 non-null   str    
 10  user_name            1464 non-null   str    
 11  review_id            1464 non-null   str    
 12  review_title         1464 non-null   str    
 13  review_content       1464 non-null   str    
 14  img_link             1464 non-null   str    
 15  product_link         1464 non-null   str 

## 4. Descriptive Statistics

In [5]:
df_clean[['discounted_price','actual_price','discount_pct','rating','rating_count','savings']].describe().round(2)


,discounted_price,actual_price,discount_pct,rating,rating_count,savings
count,1464.00,1464.00,1464.00,1464.00,1462.00,1464.00
mean,3126.01,5447.00,47.71,4.10,18307.38,2320.99
std,6946.63,10878.27,21.63,0.29,42766.10,4605.77
min,39.00,39.00,0.00,2.00,2.00,0.00
25%,323.75,800.00,32.00,4.00,1191.50,370.75
50%,799.00,1650.00,50.00,4.10,5179.00,800.00
75%,1999.00,4303.75,63.00,4.30,17342.25,1955.00
max,77990.00,139900.00,94.00,5.00,426973.00,61910.00


## 5. Missing Value Analysis

In [6]:
missing = df_clean.isnull().sum()
missing_pct = (missing / len(df_clean) * 100).round(2)
missing_df = pd.DataFrame({'Missing Count': missing, 'Missing %': missing_pct})
missing_df = missing_df[missing_df['Missing Count'] > 0]
if missing_df.empty:
    print("✅ No missing values in cleaned dataset!")
else:
    print(missing_df)


              Missing Count  Missing %
rating_count              2       0.14


## 6. Category-wise Product Distribution

In [7]:
cat_counts = df_clean['main_category'].value_counts()
print("Products per Main Category:")
print(cat_counts.to_string())
print(f"\nTotal unique main categories : {df_clean['main_category'].nunique()}")
print(f"Total unique sub-categories  : {df_clean['sub_category'].nunique()}")


Products per Main Category:
main_category
Electronics              526
Computers&Accessories    453
Home&Kitchen             447
OfficeProducts            31
MusicalInstruments         2
HomeImprovement            2
Toys&Games                 1
Car&Motorbike              1
Health&PersonalCare        1

Total unique main categories : 9
Total unique sub-categories  : 29


## 7. Rating Distribution & Anomalies

In [8]:
print("Rating Distribution:")
print(df_clean['rating'].describe().round(3))
print()
print("Products rated below 3.0 (poor) :", (df_clean['rating'] < 3.0).sum())
print("Products rated 4.0 and above    :", (df_clean['rating'] >= 4.0).sum())
print("Products rated 4.5 and above    :", (df_clean['rating'] >= 4.5).sum())


Rating Distribution:
count    1464.000
mean        4.097
std         0.292
min         2.000
25%         4.000
50%         4.100
75%         4.300
max         5.000
Name: rating, dtype: float64

Products rated below 3.0 (poor) : 6
Products rated 4.0 and above    : 1110
Products rated 4.5 and above    : 104


## 8. Discount & Price Analysis

In [9]:
print("Discount % Distribution:")
print(df_clean['discount_pct'].describe().round(2))
print()
print("Products with >70% discount :", (df_clean['discount_pct'] > 70).sum())
print("Products with <20% discount :", (df_clean['discount_pct'] < 20).sum())
print()
print("Highest savings product:")
top_saving = df_clean.nlargest(1, 'savings')[['product_name','actual_price','discounted_price','savings']]
print(top_saving.to_string(index=False))


Discount % Distribution:
count    1464.00
mean       47.71
std        21.63
min         0.00
25%        32.00
50%        50.00
75%        63.00
max        94.00
Name: discount_pct, dtype: float64

Products with >70% discount : 219
Products with <20% discount : 165

Highest savings product:
                                                                    product_name  actual_price  discounted_price  savings
Sony Bravia 164 cm (65 inches) 4K Ultra HD Smart LED Google TV KD-65X74K (Black)      139900.0           77990.0  61910.0


## 9. Correlation Analysis

In [10]:
corr_cols = ['discounted_price','actual_price','discount_pct','rating','rating_count','savings']
corr_matrix = df_clean[corr_cols].corr().round(3)
print("Correlation Matrix:")
print(corr_matrix)
r = corr_matrix.loc['discount_pct','rating']
print(f"\n🔍 Discount % vs Rating correlation: {r:.3f}")


Correlation Matrix:
                  discounted_price  actual_price  discount_pct  rating  \
discounted_price             1.000         0.962        -0.243   0.120   
actual_price                 0.962         1.000        -0.118   0.122   
discount_pct                -0.243        -0.118         1.000  -0.155   
rating                       0.120         0.122        -0.155   1.000   
rating_count                -0.027        -0.036         0.011   0.102   
savings                      0.764         0.911         0.086   0.106   

                  rating_count  savings  
discounted_price        -0.027    0.764  
actual_price            -0.036    0.911  
discount_pct             0.011    0.086  
rating                   0.102    0.106  
rating_count             1.000   -0.044  
savings                 -0.044    1.000  

🔍 Discount % vs Rating correlation: -0.155


## 10. Hypothesis — Does Higher Discount = Lower Rating?

In [11]:
df_clean['discount_bucket'] = pd.cut(df_clean['discount_pct'],
    bins=[0, 20, 40, 60, 80, 100],
    labels=['0-20%','21-40%','41-60%','61-80%','81-100%'])

bucket_rating = df_clean.groupby('discount_bucket', observed=True)['rating'].agg(['mean','count']).round(3)
bucket_rating.columns = ['Avg Rating', 'Product Count']
print("Avg Rating by Discount Bucket:")
print(bucket_rating.to_string())
print()
print("💡 Insight: Very high discount products (80-100%) show slightly lower avg ratings.")


Avg Rating by Discount Bucket:
                 Avg Rating  Product Count
discount_bucket                           
0-20%                 4.147            130
21-40%                4.126            338
41-60%                4.079            508
61-80%                4.076            380
81-100%               3.976             59

💡 Insight: Very high discount products (80-100%) show slightly lower avg ratings.


---
## 📌 EDA Summary

| # | Finding |
|---|---------|
| 1 | 1,462 valid products after removing 3 corrupt rating entries |
| 2 | Electronics is the dominant category (~36%) |
| 3 | Average rating is ~4.1 — products are generally well-received |
| 4 | Average discount is ~47% — Amazon India heavily discounts products |
| 5 | Discount % and Rating have negligible correlation |
| 6 | Higher MRP products offer bigger absolute savings |

---
*CodeAlpha Data Analytics Internship | Task 2*
